apollo-router/src/plugins/response_cache/cache_key.rs

```diff
@@ -0,0 +1,202 @@
+            let _ = write!(&mut key, ":{private_id}");
+        }
+
+        key
+    }
+}
+
+/// Cache key for an entity
+pub(super) struct PrimaryCacheKeyEntity<'a> {
+    pub(super) subgraph_name: &'a str,
+    pub(super) entity_type: &'a str,
+    pub(super) representation: &'a mut Map<ByteString, Value>,
+    pub(super) entity_key: &'a Map<ByteString, Value>,
+    /// Already hashed to to re-hash for every entities
+    pub(super) subgraph_query_hash: &'a str,
+    pub(super) additional_data_hash: &'a str,
+    pub(super) private_id: Option<&'a str>,
+}
+
+impl<'a> PrimaryCacheKeyEntity<'a> {
+    pub(super) fn hash(&mut self) -> String {
+        let Self {
+            subgraph_name,
+            entity_type,
+            subgraph_query_hash,
+            additional_data_hash,
+            private_id,
+            representation,
+            entity_key,
+        } = self;
+
+        let hashed_representation = if representation.is_empty() {
+            String::new()
+        } else {
+            hash_other_representation(representation)
+        };
+        let hashed_entity_key = hash_entity_key(entity_key);
+
+        // - response cache version: current version of the hash
+        // - subgraph name: caching is done per subgraph
+        // - type: can invalidate all instances of a type
+        // - entity key: invalidate a specific entity
+        // - representation: representation variable value
+        // - query hash: invalidate the entry for a specific query and operation name
+        // - additional data: separate cache entries depending on info like authorization status
+        let mut key = format!(
+            "version:{RESPONSE_CACHE_VERSION}:subgraph:{subgraph_name}:type:{entity_type}:entity:{hashed_entity_key}:representation:{hashed_representation}:hash:{subgraph_query_hash}:data:{additional_data_hash}"
+        );
+
+        if let Some(private_id) = private_id {
+            let _ = write!(&mut key, ":{private_id}");
+        }
+
+        key
+    }
+}
+
+/// Hash subgraph query
+pub(super) fn hash_query(query_hash: &QueryHash) -> String {
+    let mut digest = blake3::Hasher::new();
+    digest.update(query_hash.as_bytes());
+    digest.update(&[0u8; 1][..]);
+
+    digest.finalize().to_hex().to_string()
+}
+
+pub(super) fn hash_additional_data(
+    body: &mut graphql::Request,
+    context: &Context,
+    cache_key: &CacheKeyMetadata,
+) -> String {
+    let mut hasher = blake3::Hasher::new();
+
+    let repr_key = ByteString::from(REPRESENTATIONS);
+    // Removing the representations variable because it's already part of the cache key
+    let representations = body.variables.remove(&repr_key);
+    body.variables.sort_keys();
+    body.variables
+        .serialize(Blake3Serializer::new(&mut hasher))
+        .expect("this serializer doesn't throw any errors; qed");
+    if let Some(representations) = representations {
+        body.variables.insert(repr_key, representations);
+    }
```

Comment:
This being a mutable reference feels kind of dangerous to me - I wouldn't expect a hash function to mutate the body?

It also feels like it might be simpler to not have the `Blake3Serializer` and instead update the hash incrementally, ie something like:
```suggestion
pub(super) fn hash_additional_data(
    body: &graphql::Request,
    context: &Context,
    cache_key: &CacheKeyMetadata,
) -> String {
    let mut hasher = blake3::Hasher::new();

    let repr_key = ByteString::from(REPRESENTATIONS);
    for (key, value) in body.variables.iter().sorted_by(|a, b| a.0.cmp(b.0)).filter(|key| key != repr_key) {
        hasher.update(key);
        hasher.update(value);
    }
```

That might not be a good idea, just wanted to throw it out since `Blake3Serializer` is only used in this function.

Reply:
done